java/arcs/core/util/MutableBiMap.kt

```diff
@@ -0,0 +1,129 @@
+package arcs.core.util
+
+import kotlinx.coroutines.sync.Mutex
+
+/**
+ * Thread-safe mutable bi-directional map utility class.
+ */
+class MutableBiMap<L, R>() {
+    private val left2right: MutableMap<L, R> = mutableMapOf()
+    private val right2left: MutableMap<R, L> = mutableMapOf()
+    private val mutex: Mutex = Mutex()
+
+    fun put(left: L, right: R) {
+        while (!mutex.tryLock()) { /* Wait. */ }
+        if (left2right.contains(left)) {
+            right2left.remove(left2right.get(left))
+        }
+        if (right2left.contains(right)) {
+            left2right.remove(right2left.get(right))
+        }
+        left2right.put(left, right)
+        right2left.put(right, left)
+        mutex.unlock()
+    }
+
+    fun getL(right: R): L? {
+        while (!mutex.tryLock()) { /* Wait. */ }
+        val left = right2left.get(right)
+        mutex.unlock()
+        return left
+    }
+
+    fun getR(left: L): R? {
+        while (!mutex.tryLock()) { /* Wait. */ }
+        val right = left2right.get(left)
+        mutex.unlock()
+        return right
+    }
+
+    fun containsL(left: L): Boolean {
+        while (!mutex.tryLock()) { /* Wait. */ }
+        val hasLeft = left2right.contains(left)
+        mutex.unlock()
+        return hasLeft
+    }
+
+    fun containsR(right: R): Boolean {
+        while (!mutex.tryLock()) { /* Wait. */ }
+        val hasRight = right2left.contains(right)
+        mutex.unlock()
+        return hasRight
+    }
+
+    fun remove(left: L, right: R): Boolean {
+        while (!mutex.tryLock()) { /* Wait. */ }
+        var successfulRemove = false
+
+        // Ensure left is mapped to right
+        val actualRight = left2right.get(left)
+        val actualLeft = right2left.get(right)
+        if (actualRight == right && actualLeft == left) {
+            successfulRemove = left2right.remove(left, right) && right2left.remove(right, left)
+        }
+
+        mutex.unlock()
+        return successfulRemove
+    }
+
+    fun removeL(left: L): R? {
+        while (!mutex.tryLock()) { /* Wait. */ }
+        val right = left2right.remove(left)
+        right2left.remove(right, left)
+        mutex.unlock()
+        return right
+    }
+
+    fun removeR(right: R): L? {
+        while (!mutex.tryLock()) { /* Wait. */ }
+        val left = right2left.remove(right)
+        left2right.remove(left, right)
+        mutex.unlock()
+        return left
+    }
+
+    fun clear() {
+        while (!mutex.tryLock()) { /* Wait. */ }
+        left2right.clear()
+        right2left.clear()
+        mutex.unlock()
+    }
+
+    fun entries(): MutableSet<MutableMap.MutableEntry<L, R>> {
+        while (!mutex.tryLock()) { /* Wait. */ }
+        val left2rightEntries = left2right.entries
+        mutex.unlock()
+        return left2rightEntries
+    }
+
+    fun lefts(): MutableSet<L> {
+        while (!mutex.tryLock()) { /* Wait. */ }
+        val leftKeys = left2right.keys
+        mutex.unlock()
+        return leftKeys
```

Comment:
This pattern is super common in this file (spinlocking on the mutex), it'd be nice to DRY things out a bit with a helper like:

```kotlin
private inline fun <T> Mutex.withLockBlocking(inner: () -> T): T {
    while (!mutex.tryLock()) { /* Wait. */ }
    val result = inner()
    mutex.unlock()
    return result
}
```

Then in places where it's used, it'd be a bit cleaner:

```kotlin
fun clear() = mutex.withLockBlocking<Unit> {
    left2right.clear()
    right2left.clear()
}
```

Reply:
I've removed this pattern and instead made this data structure not thread safe. I plan to use this data structure in the Direct Store Muxer and will handle the thread safety there.